src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -1149,23 +1149,65 @@ public override BoundNode VisitDynamicInvocation(BoundDynamicInvocation node)
             return null;
         }
 
-        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node)
+        protected BoundNode VisitInterpolatedStringBase(BoundInterpolatedStringBase node, InterpolatedStringHandlerData? data)
         {
-            foreach (var expr in node.Parts)
+            // If there can be any branching, then we need to treat the expressions
+            // as optionally evaluated. Otherwise, we treat them as always evaluated
+            switch (data)
             {
-                VisitRvalue(expr);
+                case null:
+                    visitParts();
+                    break;
+                case { HasTrailingHandlerValidityParameter: false, UsesBoolReturns: false, Construction: var construction }:
+                    VisitRvalue(construction);
+                    visitParts();
+                    break;
+                case { UsesBoolReturns: var usesBoolReturns, Construction: var construction }:
+                    VisitRvalue(construction);
+                    var beforePartsState = State.Clone();
```

Comment:
Feels like we should have different behavior depending on value of `HasTrailingHandlerValidityParameter`, but that value is ignored here.
If it is `false`, we're guaranteed to execute the first append call. If it's `true`, we may not execute any append.
The behavior as implemented seems to assume that `HasTrailingHandlerValidityParameter` was `true`.

Once we have that, we should take another look at the second case and see whether it's still necessary (I suspect it should just fall out from the last case using `false` and `false`).

Reply:
Given that the behavior in this section is predicated on there being some amount of conditional evaluation, I think it's cleaner to keep it separate.